### DISCUSSION TITLE
Embedded Nerd Font not working on window title

### DISCUSSION THREAD

**Discussion by jackplus-xyz** [ghostty-org/ghostty#4094] [Q&A] [answered]
## Description

Nerd Font symbols fail to render in the window title bar when using a non-patched font in the `window-title-font-family` setting.

- The issue is isolated to the window title bar
- Nerd Font symbols render correctly inside the terminal

## Example

<img width="518" alt="Screenshot 2024-12-30 at 09 19 31" src="https://github.com/user-attachments/assets/ff43c1f2-9d82-46a0-8021-bbfd024d445f" />

## Reproduce

1. Terminal: Ghostty

2. Font: [TX-02 Berkeley Mono™ Typeface](https://usgraphics.com/products/berkeley-mono)

```config
# $HOME/.config/ghostty/config
window-title-font-family = TX-02
```

## Attempted Solutions

✅ Successful:

- Using a Nerd Font patched font family

❌ Unsuccessful:

- Setting `title` property with Nerd Font symbols while using non-patched fonts

**Comment by suliatis** [comment]
I just started to experiment with configuring ghostty's title string and I ran into some inconsistent behavior using nerd font icons in the title. My use case is that I want configure a custom title string from Neovim to show some relevant info both in the title and the statusline (and also for vanity 😄). I was surprised that the nerd font icons rendered in the title at all, but later realized that it works inconsistently.

Here's when it works properly rendering both the lua icon for the filetype and the git branch icon:

<img width="1224" alt="Screenshot 2025-01-05 at PM 11 13 27 copy" src="https://github.com/user-attachments/assets/258e4dd1-0704-4422-b9f9-17fa0ac9f14f" />

And here is an example when it's broken, both the clojure icon and the git branch icon is missing:

<img width="1235" alt="Screenshot 2025-01-05 at PM 11 13 37" src="https://github.com/user-attachments/assets/878301c0-7e78-4a33-9fe0-1ac8d741fb76" />

What is interesting that when it is broken then all the icons are missing, even the ones are worked before.

My ghostty config:

```
# vim set ft=conf
font-size = 13
font-family = SF Mono
adjust-cell-height = 20%

macos-titlebar-proxy-icon = hidden
window-height = 40
window-width = 140
window-padding-x = 0
window-padding-y = 0,2

confirm-close-surface = false

# I use custom themes, but don't think it is relevant
# theme = dark:github_dark_default,light:github_light_default
background-opacity = 0.9
background-blur-radius = 20
```

And here is a neovim config to reproduce, using some mini.nvim plugins:

```
local path_package = vim.fn.stdpath 'data' .. '/site'
local mini_path = path_package .. '/pack/deps/start/mini.nvim'
if not vim.loop.fs_stat(mini_path) then
  vim.cmd 'echo "Installing `mini.nvim`" | redraw'
  local clone_cmd = {
    'git',
    'clone',
    '--filter=blob:none',
    'https://github.com/echasnovski/mini.nvim',
    mini_path,
  }
  vim.fn.system(clone_cmd)
  vim.cmd 'packadd mini.nvim | helptags ALL'
end

local MiniDeps = require 'mini.deps'
MiniDeps.setup { path = { package = path_package } }
local add, now, later = MiniDeps.add, MiniDeps.now, MiniDeps.later

now(function()
  local MiniGit = require 'mini.git'
  MiniGit.setup {}

  local MiniIcons = require 'mini.icons'
  MiniIcons.setup {}
  MiniIcons.mock_nvim_web_devicons()
  MiniIcons.tweak_lsp_kind()

  local M = {}

  _G.TITLESTRING_CACHE = nil

  local default_config = {
    options = {
      use_icons = true,
    },
    content = {
      {
        title = function()
          return M.combine_sections {
            '%=',
            M.section_filename {
              ambigous_short_names = {
                'init.lua',
                'config',
              },
            },
            '-',
            M.section_cwd(),
            M.section_git(),
            '%=',
          }
        end,
        filter = function(buf, _)
          local buftype = vim.bo[buf].buftype
          return buftype ~= 'nofile' and buftype ~= 'terminal'
        end,
      },
      trouble = {
        title = function()
          return nil
        end,
      },
    },
  }

  function M.setup(config)
    -- TODO: validate config
    default_config = vim.tbl_deep_extend('force', default_config, config)
    _G.Statusline = M

    vim.go.title = true
    vim.go.titlestring = '%{%v:lua.Statusline.titlestring()%}'
  end

  function M.titlestring()
    local ft = vim.bo.filetype
    local content = default_config.content[ft] or default_config.content[1]

    local win = vim.api.nvim_get_current_win()
    local buf = vim.api.nvim_get_current_buf()
    if content.filter == nil or content.filter(buf, win) then
      local title = content.title()
      if title then
        TITLESTRING_CACHE = title
        return title
      end
    end

    return TITLESTRING_CACHE
  end

  function M.combine_sections(sections)
    sections = vim.tbl_filter(function(section)
      return section ~= nil and section ~= ''
    end, sections)

    return table.concat(sections, ' ')
  end

  function M.section_filename(args)
    args = args or {}
    local path = vim.fn.expand '%:p'
    local fmt = args.format or M.format_filename_default
    return fmt(path, args)
  end

  function M.format_filename_default(path, args)
    return M.format_filename_short(path, args)
  end

  function M.format_filename_short(path, args)
    local short_name = vim.fn.fnamemodify(path, ':t')
    if args.ambigous_short_names and vim.tbl_contains(args.ambigous_short_names, short_name) then
      local parent_name = vim.fn.fnamemodify(path, ':h:t')
      short_name = parent_name .. '/' .. short_name
    end

    local file_icon = MiniIcons.get('extension', path)

    return file_icon .. ' ' .. short_name
  end

  function M.section_git()
    local summary = vim.b.minigit_summary_string or vim.b.gitsigns_head
    if not summary then
      return ''
    end

    if default_config.options.use_icons then
      local icon = ''
      return icon .. ' ' .. summary
    end

    return summary
  end

  function M.section_cwd()
    local cwd = vim.fn.getcwd()
    local parent = vim.fn.fnamemodify(cwd, ':h:t')
    local dir = vim.fn.fnamemodify(cwd, ':t')

    if parent == 'code' or dir == '.dots' then
      return dir
    end

    return cwd
  end

  function M.format_cwd_short(path)
    local short_name = vim.fn.fnamemodify(path, ':t')
    return short_name
  end

  M.setup {}
end)
```

**Comment by jcollie** [comment, marked as answer]
The embedded font only works inside the terminal window. Fonts in UI elements like menus, window titles, and tab titles can only use fonts that are installed on your system.

**Reply by suliatis** [reply]
Oh, I thought it enough to have the `Symbols Nerd Font` installed to my system. But you saying that ghostty is embedding those symbols and it's only working within the terminal. To have icons in the title I have to install and set a `window-title-font-family` to a font with nerd fonts baked in.

<img width="1112" alt="Screenshot 2025-01-06 at AM 7 50 35" src="https://github.com/user-attachments/assets/84387957-09b2-44b0-9e38-bc74ecd4d9fa" />

```
window-title-font-family = SFMono Nerd Font
```

<img width="1302" alt="Screenshot 2025-01-06 at AM 7 51 34" src="https://github.com/user-attachments/assets/c6300fde-beab-4b3e-a71c-cffbf7beb2cd" />

Now the icons are working properly, but I ran into another issue using custom font in the title. The last 3 chars of the title is always trimmed off using either `SF Mono` or `SFMono Nerd Font`. The only I could fix it is to use a font with fixed width like `Iosevka Fixed`.

**Reply by suliatis** [reply]
Also one thing is not clear to me. If embedded nerd fonts are not supported on UI elements, then how can certain icons work (like lua and git branch), while others not (like scala and clojure)?

**Comment by suliatis** [comment]
@jcollie I still don't get this, because the `Symbols Nerd Font Mono` is installed on my system. Also when the nerd font symbols are working, then all of them are shown properly: 

<img width="1224" alt="Screenshot 2025-01-05 at PM 11 13 27 copy" src="https://github.com/user-attachments/assets/22de8f8e-b496-48d3-a079-f484cbdc094f" />

But otherwise, the previously shown symbols are broken too. Like the second symbol below, that should be a git branch icon and shown properly on the previous image:

<img width="1235" alt="Screenshot 2025-01-05 at PM 11 13 37" src="https://github.com/user-attachments/assets/edb08808-100e-4d28-8e7e-3ccf704ffe4c" />
